### DISCUSSION TITLE
How to set up WW3 within UFS limited area domain configurations?

### DISCUSSION THREAD

**Discussion by XiaSun-Atmos** [NOAA-EMC/WW3#819] [Q&A]
Hi,
What files are needed for running UFS regional domain (e.g. RRFS) with coupled WW3? 
Thanks!
Xia Sun-CIRES/GSL

**Comment by JessicaMeixner-NOAA** [comment]
Hi @XiaSun-Atmos that's a great question!  We'd first need a grid that covers the region of interest that matches RRFS. Otherwise the set-up would be similar to how WW3 is set up for other regional domains such as HAFS (there are UFS regression tests for that we could follow).   What region of RRFS domain are you looking to cover with a wave grid?

**Comment by XiaSun-Atmos** [comment]
Thank you @JessicaMeixner-NOAA for your quick response! The ultimate domain would be on the east coast for WFIP3 using RRFS. Below is the domain info we made for the UFS LAMs for use in ufs srw app. 

Center Lon: -70.8
Center Lat: 40.6
Horizontal dimension X: 700
Horizontal dimension y: 600
Grid distance: 1km
I checked the hafs regression test and changed the suite to FV3_HRRR, suitable namelist, and corresponding diag/field tables for RRFS. It gave errors like missing connections of variables listed in the nems_configure file.

**Reply by JessicaMeixner-NOAA** [reply]
I don't see any issues with the logic of how you set-up that test runs and it seems like something that "should" work.   What computer are you running the test? Can you share more details about the error messages?

**Reply by XiaSun-Atmos** [reply]
I am running it on Hera. This is the err msg in the PET000.ESMF_LogFile:
```
20221014 141416.993 ERROR            PET000 atmos_model.F90:3092 Pointer must already be allocated  - Failure to populate exported field: inst_zonal_wind_height10
20221014 141416.993 ERROR            PET000 module_fcst_grid_comp.F90:463 Pointer must already be allocated  - Passing error in return code
20221014 141416.993 ERROR            PET000 module_fcst_grid_comp.F90:1136 Pointer must already be allocated  - Passing error in return code
20221014 141416.993 ERROR            PET000 fv3_cap.F90:957 Pointer must already be allocated  - Passing error in return code
20221014 141416.993 ERROR            PET000 ATM:src/addon/NUOPC/src/NUOPC_ModelBase.F90:986 Pointer must already be allocated  - Passing error in return code
20221014 141416.993 ERROR            PET000 UFS Driver Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:2577 Pointer must already be allocated  - Phase 'IPDvXp04' Initialize for modelComp 2: ATM did not return ESMF_SUCCESS
```
There are some other prints like:
`20221014 141416.993 INFO             PET000 module_cplfields.F90:519 realizeConnectedCplFields FV3 Export Field t2m is not connected`
But these 'not connected' msg also existed in successful hafsw RT run.

**Reply by JessicaMeixner-NOAA** [reply]
@XiaSun-Atmos apologies I lost track of this over the weekend.  Do you have cplwav=true in your input.nml?  Also what is your coupling mode in the mediator attributes of the nems.configure file?

**Reply by XiaSun-Atmos** [reply]
No worries at all @JessicaMeixner-NOAA . Thank you for helping me with this!
I used the nems.configure file for nems.configure.hafs_atm_wav.IN, so the coupling model is `hafs`.  
While I just found out that even though I set it up in the configure file under` /tests/tests` dir, It does not go to the` input.nml` for a reason
```
export CPLWAV=.true.
export CPLWAV2ATM=.true.
```
I added `cplwav=true` and `cplwav2atm=true` to the `input.nml `and the err msg with` pointer must be allocated` is gone now. It proceeded further than earlier, but still crashes with seg fault, the first few lines are:
```
297: ==== backtrace (tid:  98581) ====
297:  0 0x000000000004d455 ucs_debug_print_backtrace()  ???:0
297:  1 0x00000000024a1aaa ESMCI::ArrayBundle::sparseMatMul()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/src/ESMCI_ArrayBundle.C:1675
297:  2 0x0000000001561410 c_esmc_arraybundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/interface/ESMCI_ArrayBundle_F.C:717
297:  3 0x0000000000da5ed5 esmf_arraybundlemod_mp_esmf_arraybundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/interface/ESMF_ArrayBundle.F90:2945
297:  4 0x0000000000d05ece esmf_fieldbundlemod_mp_esmf_fieldbundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/FieldBundle/src/ESMF_FieldBundle.F90:16626
297:  5 0x000000000584b28e fv3gfs_cap_mod_mp_modeladvance_phase2_()  /home/Xia.Sun/wrfruc_wk/WFIP3/ufs-weather-model-oct18/FV3/fv3_cap.F90:1104
```

**Reply by JessicaMeixner-NOAA** [reply]
Glad things are going further! You might try giving more cores or threads to the job? Sometimes seg-faults are just memory issues?  Have you see any helpful info the in the PET files?

**Reply by XiaSun-Atmos** [reply]
Thank you @JessicaMeixner-NOAA for the suggestions. I played a little bit on the cores (360->480) and OMP_THREADS (from 1 ->2), but seems no luck. The err msg. still starts with:
```
  0:   ... DONE microphysical lookup tables
240: [h18c05:92717:0:92717] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
```
 Are there other memory options I might try to touch?
The model crashes at: 
```
300:  WW3 log written to /scratch2/BMC/wrfruc/xiasun/WFIP3/ufs-weather-model-oct18/st
300:  mp2/Xia.Sun/FV3_RT/rt_73593/rrfs_regional_atm_wav/./log.ww3
  0:  fcst_realize, cpl_grid_id=           1
240:   aft fcst run output time=          60 FBcount=           3 na=           1
```
The PET files ends at:
```
20221026 184441.559 INFO             PET000 (med_methods_FB_diagnose) (med_phases_prep_wav_avg) FBExpAccumWav before avg: Sa_u10m   0.000000     0.9990000E+21 0.1578420E+24    1656
20221026 184441.559 INFO             PET000 (med_methods_FB_diagnose) (med_phases_prep_wav_avg) FBExpAccumWav before avg: Sa_v10m   0.000000     0.9990000E+21 0.1578420E+24    1656
20221026 184441.559 INFO             PET000 (med_methods_FB_diagnose): done
```
The log.ww3 ends at: 
```
 Initial time     : 2019/08/29 00:00:00 UTC


                                       |         input         |      output      |
                                       |-----------------------|------------------|
     step | pass |    date      time   | b w l c t r i i1 i5 d | g p t r b f c r2 |
  --------|------|---------------------|-----------------------|------------------|
  --------+------+---------------------+---------------------------+--------------+
 mesh file for domain is mesh.hafs.nc
ymd2date currTime wav_comp_nuopc hh,mm,ss,ymd   0   0   0  20190829
```

**Reply by JessicaMeixner-NOAA** [reply]
is 240 a wave or atm model pet? what does PET log for PE 240 say? Or the last and or first pet log that's written say?

**Reply by XiaSun-Atmos** [reply]
Here are the petlist bounds from the nems.configure
```
MED_petlist_bounds:             0 239
ATM_petlist_bounds:             0 299
WAV_petlist_bounds:             300 359
```
I think 240 belongs to ATM?
PE 240 ends as:
```
20221028 182855.433 INFO             PET240 af FieldBundleRegridStore
20221028 182855.433 INFO             PET240 ... returned from wrtFB(03,01) FieldBundleRegridStore().
20221028 182857.466 INFO             PET240 Driver is in ModifyCplLists()
20221028 182857.466 INFO             PET240  Found            4  Connectors. Modifying CplList Attribute....
20221028 182907.366 DEBUG            PET240 about to destroy Grid: 0x2c0f7160
20221028 182909.811 DEBUG            PET240 about to destroy Grid: 0x2bd961a0
20221028 182947.151 INFO             PET240 fv3_checkimport wave_z0_roughness_length   1  17  F
20221028 182947.151 INFO             PET240 entering FV3_ADVANCE phase1 with clock current: 2019  8 29  0  0  0   0
20221028 182947.151 INFO             PET240 entering FV3_ADVANCE phase1 with clock start:   2019  8 29  0  0  0   0
```

The last one is PE 359:
```
20221028 171448.584 INFO             PET359 at import: Sa_u10m    0.000000     0.9990000E+21 0.1005993E+25    3311
20221028 171448.584 INFO             PET359 at import: Sa_v10m    0.000000     0.9990000E+21 0.1005993E+25    3311
20221028 171448.585 INFO             PET359 (wav_import_export:set_importmask) called
20221028 171448.587 INFO             PET359 Setting initial import_mask at currTime : 2019  8 29  0  0  0   0
20221028 171448.587 INFO             PET359 (wav_import_export:set_importmask) done

```
The first PE 000 ends:
```
20221028 171448.764 INFO             PET000  (Set Run Clock)  done
20221028 171448.764 INFO             PET000 (med_methods_FB_diagnose): called
20221028 171448.764 INFO             PET000 (med_methods_FB_diagnose) (med_phases_prep_wav_avg) FBExpAccumWav before avg: Sa_u10m   0.000000     0.9990000E+21 0.8691300E+23    1656
20221028 171448.764 INFO             PET000 (med_methods_FB_diagnose) (med_phases_prep_wav_avg) FBExpAccumWav before avg: Sa_v10m   0.000000     0.9990000E+21 0.8691300E+23    1656
20221028 171448.764 INFO             PET000 (med_methods_FB_diagnose): done
```
I also tried reducing the time steps dt_atmos, which seems have no impact on this.

**Reply by JessicaMeixner-NOAA** [reply]
Anything in the PE 300?

**Reply by XiaSun-Atmos** [reply]
It ends here,:
```
20221028 185446.017 INFO             PET300 ------>Advancing WAV from: 2019  8 29  0  0  0   0
20221028 185446.017 INFO             PET300 --------------------------------> to: 2019  8 29  0  1 30   0
20221028 185446.018 INFO             PET300 (wav_import_export:import_fields) called
20221028 185446.018 INFO             PET300 at import: Sa_u10m    0.000000     0.9990000E+21 0.2016981E+25    6624
20221028 185446.018 INFO             PET300 at import: Sa_v10m    0.000000     0.9990000E+21 0.2016981E+25    6624
20221028 185446.020 INFO             PET300 (wav_import_export:set_importmask) called
20221028 185446.022 INFO             PET300 Setting initial import_mask at currTime : 2019  8 29  0  0  0   0
20221028 185446.022 INFO             PET300 (wav_import_export:set_importmask) done
```

**Reply by JessicaMeixner-NOAA** [reply]
None of this is pointing to where the segfualt is.....

**Reply by XiaSun-Atmos** [reply]
I agree. It seems it stops at model advance stage but no specifics. The only info I can see is the stack trace at the` err` file under the msg of seg fault. But I am not sure how useful it is.
```
281: ==== backtrace (tid: 296246) ====
281:  0 0x000000000004d455 ucs_debug_print_backtrace()  ???:0
281:  1 0x00000000024a1aaa ESMCI::ArrayBundle::sparseMatMul()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/src/ESMCI_ArrayBundle.C:1675
281:  2 0x0000000001561410 c_esmc_arraybundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/interface/ESMCI_ArrayBundle_F.C:717
281:  3 0x0000000000da5ed5 esmf_arraybundlemod_mp_esmf_arraybundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/ArrayBundle/interface/ESMF_ArrayBundle.F90:2945
281:  4 0x0000000000d05ece esmf_fieldbundlemod_mp_esmf_fieldbundlesmm_()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Infrastructure/FieldBundle/src/ESMF_FieldBundle.F90:16626
281:  5 0x000000000584b28e fv3gfs_cap_mod_mp_modeladvance_phase2_()  /home/Xia.Sun/wrfruc_wk/WFIP3/ufs-weather-model-oct18/FV3/fv3_cap.F90:1104
281:  6 0x00000000058471ad fv3gfs_cap_mod_mp_modeladvance_()  /home/Xia.Sun/wrfruc_wk/WFIP3/ufs-weather-model-oct18/FV3/fv3_cap.F90:977
281:  7 0x00000000022bd916 ESMCI::MethodElement::execute()  /scratch2/NCEPDEV/nwprod/hpc-stack/src/v1.2.0/pkg/v8.3.0b09/src/Superstructure/Component/src/ESMCI_MethodTable.C:377
```

**Reply by XiaSun-Atmos** [reply]
@JessicaMeixner-NOAA Update: It turns out this err msg also existed when using FV3_HRRR suite without wave turned on for the hafs RT. I guess there is some setup incompatible with FV3_HRRR. Since I am not familiar with hafs, I managed to start a new successful test based on rrfs RT but with the domain from hafs RT (i.e., grids setup and input) without wave. 

 If I turned wave on  with `nems.configure.hafs_atm_wav.IN` , `MED_model=cmeps` (`export_hafs_regional`), and coupling model of `hafs`, it gives err of application called MPI_Abort and in the PET000 file shows:
`20221102 224340.863 ERROR            PET000 UFSDriver.F90:519 Not valid  -  No component cmeps found`

Did I miss something? My question is for the regional setup, do we must use `MED_model `and `CPLMODE=hafs`? Which entries are necessary in the nems.configure file for regional setup?
Thanks!

**Reply by JessicaMeixner-NOAA** [reply]
Depends on which application you are compiling.  You could follow a WAV-ATM test with CMEPS and use the mesh cap (which is used in HAFS and S2S), which would require to follow something like what is in the PR here: https://github.com/ufs-community/ufs-weather-model/pull/1407/files for the added test  atmwav_control_noaero_p8. (It's my understanding that there is not the flexibility with the mesh cap to not use the mediator at this time). 

Or, you can use the wmesmf cap which allows for direct connectors and you can follow the set-up in the test for app ATMW using the test control_atmwav.

**Reply by XiaSun-Atmos** [reply]
You are right @JessicaMeixner-NOAA ! Once I made sure the configuration works for FV3_HRRR without wave, following with the suitable app (HAFSW), it's generating forecasts now. Thank you so much!
If I would like to change a domain, which files should I touch for WW3 input and WW3 namelist?
Thanks!

**Reply by JessicaMeixner-NOAA** [reply]
@XiaSun-Atmos depending on what you'd want for your wave domain, you'd likely have to generate a new grid to match what you'd want for your different domain, and then you'd need to make a new mod_def file to match that.  There's a software here: https://github.com/noaa-emc/gridgen that helps with structured grid generation.  So glad things are running!

**Reply by XiaSun-Atmos** [reply]
Thanks @JessicaMeixner-NOAA ! I downloaded the gridgen and generated 4 outputs: `depth_ascii, maskorig_ascii, meta, obstr_lev1`
To create a ww3_grid.inp/nml, do you have a suggested reg test in WW3 I can start with for a regional domain?
Thanks!

**Reply by JessicaMeixner-NOAA** [reply]
@XiaSun-Atmos there are several that use regional domains, maybe try tp2.4?

**Reply by XiaSun-Atmos** [reply]
Thanks @JessicaMeixner-NOAA . tp2.4 provides good examples for my test. I started a test for rrfs domain. The `mesh_rrfs.nc `file is made using the `scrip.nc `file generated from `ww3_grid` and command of `ESMF_Scrip2Unstruct`. I feed` ww3_grid` with `lat.dat `and `lon.dat` from rrfs with 3-km domain.
When running the model, the err msg is:
```
511:  *** WAVEWATCH III ERROR IN W3FLDO :
511:      INCOMPATIBLE GRID DATA :         901        441          1
511:                   SHOULD BE :        1799       1059          2
```
Does the grid data here mean `wind.ww3` (`wind.natl_6m`) which I copied from  `@[INPUTDATA_ROOT_WW3]/wind.natl_6m ./wind.ww3`?
Should I prepare `wind.ww3` using `mod_def.ww3` corresponding to rrfs domain? Thanks!

**Reply by JessicaMeixner-NOAA** [reply]
The wind.natl_6m  is specific to the corresponding 'natl_6m' grid.  You will have to regenerate wind.ww3 data for your coupled model if you'd like to use wind input from another source that's not covered by your atm model domain in the coupled model or if you're running WW3 standalone.

**Reply by XiaSun-Atmos** [reply]
@JessicaMeixner-NOAA , I tried to make the wind.ww3 file for my domain. The model crashes with this err:
```
  0:   ... DONE microphysical lookup tables
184: [h4c35:215431:0:215431] Caught signal 8 (Floating point exception: integer divide by zero)
```
I ran WW3 offline, it runs and generates outputs.
My ww3_grid.nml is attached. Did you spot anything not right? Also, how should I define the &INBND_POINT_NML? 
Thank you so much!
```
! -------------------------------------------------------------------- !
! WAVEWATCH III - ww3_grid.nml - Grid pre-processing                   !
! -------------------------------------------------------------------- !

! -------------------------------------------------------------------- !
! Define the spectrum parameterization via SPECTRUM_NML namelist
! -------------------------------------------------------------------- !
&SPECTRUM_NML
  SPECTRUM%XFR       =  1.1
  SPECTRUM%FREQ1     =  0.0373
  SPECTRUM%NK        =  32
  SPECTRUM%NTH       =  24
/

! -------------------------------------------------------------------- !
! Define the run parameterization via RUN_NML namelist
! -------------------------------------------------------------------- !
&RUN_NML
  RUN%FLCX         =  T
  RUN%FLCY         =  T
  RUN%FLCTH        =  T
  RUN%FLCK         =  T
/

! -------------------------------------------------------------------- !
! Define the timesteps parameterization via TIMESTEPS_NML namelist
! -------------------------------------------------------------------- !
&TIMESTEPS_NML
  TIMESTEPS%DTMAX        =  120.
  TIMESTEPS%DTXY         =  40.
  TIMESTEPS%DTKTH        =  60.
  TIMESTEPS%DTMIN        =  10.
/

! -------------------------------------------------------------------- !
! Define the grid to preprocess via GRID_NML namelist
! -------------------------------------------------------------------- !
&GRID_NML
  GRID%NAME         =  'RRFS'
  GRID%NML          =  'namelists_curv.nml'
  GRID%TYPE         =  'CURV'
  GRID%COORD        =  'SPHE'
  GRID%CLOS         =  'NONE'
  GRID%ZLIM         =  0.0
  GRID%DMIN         =  1.0
/

! -------------------------------------------------------------------- !
! Define the rectilinear grid type via CURV_NML namelist
! -------------------------------------------------------------------- !
&CURV_NML
  CURV%NX              =  1799
  CURV%NY              =  1059
  CURV%XCOORD%SF        =  1.0
  CURV%XCOORD%OFF       =  0.0
  CURV%XCOORD%FILENAME  =  'lon.1799x1059.RRFS.dat'
  CURV%YCOORD%SF        =  1.0
  CURV%YCOORD%OFF       =  0.0
  CURV%YCOORD%FILENAME  =  'lat.1799x1059.RRFS.dat'
/

! -------------------------------------------------------------------- !
! Define the depth to preprocess via DEPTH_NML namelist
! -------------------------------------------------------------------- !
&DEPTH_NML
  DEPTH%SF        =  -1.
  DEPTH%FILENAME  =  'rrfs3km.depth_ascii'
/

! -------------------------------------------------------------------- !
! Define the input boundary points via INBND_COUNT_NML and
!                                      INBND_POINT_NML namelist
! -------------------------------------------------------------------- !
&INBND_COUNT_NML
  INBND_COUNT%N_POINT    =  6
/

&INBND_POINT_NML
  INBND_POINT(1)         =  102 86 F
  INBND_POINT(2)         =  102 118 T
  INBND_POINT(3)         =  2 118 T
  INBND_POINT(4)         =  2 2 T
  INBND_POINT(5)         =  102 2 T
  INBND_POINT(6)         =  102 11 T
/

! -------------------------------------------------------------------- !
! WAVEWATCH III - end of namelist                                      !
! -------------------------------------------------------------------- !
```

**Reply by JessicaMeixner-NOAA** [reply]
I don't see any red flags with this.
